### ISSUE TITLE
RavenDB Server process not restarted

### ISSUE THREAD

**Issue by ramonsmits** [Particular/ServiceControl#3522]
#### Description

When the RavenDB5 server process is gone. Either because it crashed or was killed SC isn't detecting this state.

#### Expected behavior

SC having a recovery mechanism when the RavenDB process dies.

#### Actual behavior

Nothing happens

#### Versions

4.30.0

### Steps to reproduce

1. Start audit instance with RavenDB 5 engine
2. Kill the sub process that hosts the RavenDB 5 server
3. Observe the logs and notice nothing is logged
4. Observe the RavenDB 5 server process isn't restarted

### Relevant log output

A log file from a customer shows the log flooded with the following error:

```txt
System.Net.WebException: Unable to connect to the remote server ---> System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 127.0.0.1:44445
   at System.Net.Sockets.Socket.InternalEndConnect(IAsyncResult asyncResult)
   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean connectFailure, Socket s4, Socket s6, Socket& socket, IPAddress& address, ConnectSocketState state, IAsyncResult asyncResult, Exception& exception)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.EndGetRequestStream(IAsyncResult asyncResult, TransportContext& context)
   at System.Net.Http.HttpClientHandler.GetRequestStreamCallback(IAsyncResult ar)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Raven.Client.Http.RequestExecutor.<SendRequestToServer>d__116`1.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54038\src\Raven.Client\Http\RequestExecutor.cs:line 1039
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Http.RequestExecutor.<ExecuteAsync>d__114`1.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54038\src\Raven.Client\Http\RequestExecutor.cs:line 866
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Raven.Client.Documents.Session.AsyncDocumentSession.<SaveChangesAsync>d__32.MoveNext() in C:\Builds\RavenDB-Stable-5.4\54038\src\Raven.Client\Documents\Session\AsyncDocumentSession.cs:line 185
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ServiceControl.Audit.Persistence.RavenDb.RavenDbFailedAuditStorage.<Store>d__1.MoveNext() in /_/src/ServiceControl.Audit.Persistence.RavenDb5/RavenDbFailedAuditStorage.cs:line 27
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at CallSite.Target(Closure , CallSite , Object )
   at ServiceControl.Audit.Auditing.AuditIngestionFaultPolicy.<DoLogging>d__8.MoveNext() in /_/src/ServiceControl.Audit/Auditing/AuditIngestionFaultPolicy.cs:line 77
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at CallSite.Target(Closure , CallSite , Object )
   at ServiceControl.Audit.Auditing.AuditIngestionFaultPolicy.<Handle>d__7.MoveNext() in /_/src/ServiceControl.Audit/Auditing/AuditIngestionFaultPolicy.cs:line 59
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ServiceControl.Audit.Auditing.AuditIngestionFaultPolicy.<OnError>d__5.MoveNext() in /_/src/ServiceControl.Audit/Auditing/AuditIngestionFaultPolicy.cs:line 40
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.Msmq.ReceiveStrategy.<HandleError>d__8.MoveNext() in /_/src/NServiceBus.Transport.Msmq/ReceiveStrategy.cs:line 131
```

#### Workarounds

None

**Comment by ramonsmits** [comment]
Analysing the source used in 4.30.0 shows that the `ServerProcessExited` event isn't subscribed to

https://github.com/Particular/ServiceControl/blob/14d2548bf5eade95e5ec8042f4030b3a6ae423a5/src/ServiceControl.Audit.Persistence.RavenDb5/EmbeddedDatabase.cs

RavenDB docs ServerProcessExited event: https://ravendb.net/docs/article-page/5.3/csharp/server/embedded#serverprocessexited-event

It seems that this event must be subscribed so that a compensating action can be executed. The most logical step for me would be terminating the SC instance as this will not be a recoverable problem if the environment is no longer healthy.

**Comment by ramonsmits** [comment]
The mentioned log entry for connectivity errors should be handled gracefully and NOT dump the exception information in the log.

**Comment by jpalac** [comment]
Fixed in #3590
